src/magic_folder/magic_folder.py

```diff
@@ -2134,3 +2190,137 @@ def trap_conflicts(f):
             return False
         d.addErrback(trap_conflicts)
         return d.addActionFinish()
+
+
+@attr.s
+class LocalSnapshotCreator(object):
+    """
+    When given the db and the author instance, this class that actually
+    creates a local snapshot and stores it in the database.
+    """
+    db = attr.ib()  # our database
+    author = attr.ib(validator=attr.validators.instance_of(LocalAuthor))  # LocalAuthor instance
+    stash_dir = attr.ib(validator=attr.validators.instance_of(FilePath))
+
+    @eliotutil.inline_callbacks
+    def process_item(self, path):
+        """
+        Convert `path` into a LocalSnapshot and persist it to disk.
+
+        :param FilePath path: a single file inside our magic-folder dir
+        """
+
+        with path.open('rb') as input_stream:
+            # Query the db to check if there is an existing local
+            # snapshot for the file being added.
+            # If so, we use that as the parent.
+            mangled_name = magicpath.mangle_path(path)
+            parent_snapshot = self.db.get_local_snapshot(mangled_name, self.author)
+            if parent_snapshot is None:
+                parents = []
+            else:
+                parents = [parent_snapshot]
+
+            # need to handle remote-parents when we have remote
+            # snapshots
+
+            # when we handle conflicts we will have to handle multiple
+            # parents here (or, somewhere)
+
+            relpath_u = path.asTextMode(encoding="utf-8").path
+            SNAPSHOT_CREATOR_PROCESS_ITEM.log(relpath=relpath_u)
```

Comment:
This would most likely make more sense as an action than as a message.  That way it could contain the `create_snapshot` call and anything it logs (or its failure).